### PR TITLE
[corlib] Math from CoreFX

### PR DESCRIPTION
**Pull request by EgorBo**
Replaces reference source implementation with CoreFX
(Math.CoreCLR is copied from coreclr repo).
Also +2k math tests from CoreFX.
There are a couple of differences, e.g.:
```csharp
Console.WriteLine(Math.Pow(double.NaN, 0));
Console.WriteLine(Math.Pow(1, double.NaN));
Console.WriteLine(Math.Pow(-1, double.PositiveInfinity));
Console.WriteLine(Math.Pow(-1, double.NegativeInfinity));
```
All print `1` on .NET Core but `NaN` on .NET 4.7 and old mono (before these changes).
But @tannergooding proved it's correct behavior https://gitter.im/dotnet/corefx?at=5b6e072d5762ee4fe55bc313 (IEEE 754:2008 spec - 9.2.1: Special Values)

`Math.Round` is now managed, but it has `[Intrinsics]` so I guess it's implemented using SSE41.
Also, fixes https://github.com/mono/mono/issues/9989

Part of #7737

### CODE REVIEW

**Comment by tannergooding** [comment]
> Math.Round is now managed, but it has [Intrinsics] so I guess it's implemented using SSE41.

Yes, `Math.Round`, `Math.Ceiling`, and `Math.Floor` all use SSE4.1, on supported hardware.

**Comment by EgorBo** [comment]
@monojenkins build failed

File: mono/metadata/icall-def.h [hunk 1]
```diff
@@ -400,28 +400,32 @@ ICALL_TYPE(IOSELECTOR, "System.IOSelector", IOSELECTOR_1)
...
-NOHANDLES(ICALL(MATH_10, "Log10", ves_icall_System_Math_Log10))
-NOHANDLES(ICALL(MATH_11, "Pow", ves_icall_System_Math_Pow))
-NOHANDLES(ICALL(MATH_12, "Round", ves_icall_System_Math_Round))
-NOHANDLES(ICALL(MATH_14, "Sin", ves_icall_System_Math_Sin))
-NOHANDLES(ICALL(MATH_15, "Sinh", ves_icall_System_Math_Sinh))
-NOHANDLES(ICALL(MATH_22, "SplitFractionDouble", ves_icall_System_Math_SplitFractionDouble))
-NOHANDLES(ICALL(MATH_16, "Sqrt", ves_icall_System_Math_Sqrt))
-NOHANDLES(ICALL(MATH_17, "Tan", ves_icall_System_Math_Tan))
-NOHANDLES(ICALL(MATH_18, "Tanh", ves_icall_System_Math_Tanh))
+ICALL_TYPE(MATH, "System.Math", MATH_1)
```
Comment (kumpera) [inline mono/metadata/icall-def.h]: This is unwanted line damage

File: mono/metadata/sysmath.c [hunk 2]
```diff
@@ -108,27 +54,27 @@ ves_icall_System_Math_Sin (gdouble x)
 }
 
 gdouble 
-ves_icall_System_Math_Cos (gdouble x)
+ves_icall_System_Math_Cbrt (gdouble x)
```
Comment (kumpera) [inline mono/metadata/sysmath.c]: did you sort the functions by name?

**Comment by kumpera** [comment]
`Math.Round is now managed, but it has [Intrinsics] so I guess it's implemented using SSE41.`
Since mono is not CoreCLR, it's not. Plus, SSE41 is X86 only. This PR is a perf regression.

File: mono/metadata/sysmath.c (hunk 2 above)
Comment (EgorBo) [reply mono/metadata/sysmath.c]: um.. yes, the list looks sorted (also no warrnings about the sort order)

**Comment by EgorBo** [comment]
@kumpera intrinsics are implemented here https://github.com/mono/mono/pull/10039
but I guess yes we should keep this method native.

File: mono/metadata/icall-def.h (hunk 1 above)
Comment (EgorBo) [reply mono/metadata/icall-def.h]: could you please clarify why?

Comment (jaykrell) [reply mono/metadata/icall-def.h]: I believe @kumpera wants a smaller edit here. If you are inserting functions, you can extend the numbering scheme, e.g. insert "17a" between "17" and "18".

**Comment by EgorBo** [comment]
a small benchmark:
image
Abs is super fast because it was not inline-friendly in ref-source implementation.

**Comment by EgorBo** [comment]
@monojenkins build failed

**Comment by kumpera** [comment]
@EgorBo `[Intrinsic]` is a custom attribute ignored by mono. You can have an icall as fallback for lack of an intrinsic.

File: mono/metadata/sysmath.c (hunk 2 above)
Comment (kumpera) [reply mono/metadata/sysmath.c]: FYI, we don't do this sort of code shuffling in runtime code. We favor smaller diffs.

File: mono/metadata/icall-def.h (hunk 1 above)
Comment (kumpera) [reply mono/metadata/icall-def.h]: Jay is on spot. Larger changes that make it hard to review and reduce the value of history for no  good reason are to be avoided.

**Comment by EgorBo** [comment]
@kumpera yeah I understand that it is just a hint for user that a method is intrinsics (btw, for some reason `Math.Ceiling` and `Math.Floor` are not marked with it in CoreFX). Ok, since we can have an intrinsic on top of icall I guess it makes sense to rollback it.

File: mono/metadata/icall-def.h (hunk 1 above)
Comment (EgorBo) [reply mono/metadata/icall-def.h]: got it, now the diff is smaller, but looks a bit weird :)

File: mono/metadata/sysmath.c (hunk 2 above)
Comment (EgorBo) [reply mono/metadata/sysmath.c]: ah, sorry, I thought you mean icall-defs order. Re-order in this file was made by a mistake

**Comment by EgorBo** [comment]
@monojenkins build failed

**Comment by EgorBo** [comment]
@monojenkins build failed

**Comment by EgorBo** [comment]
@monojenkins build OS X x64 iOS SDK

File: mcs/class/corlib/Test/System/MathTest.cs [hunk 3]
```diff
@@ -499,15 +499,15 @@ public void TestPow ()
 			Assert.IsTrue (double.IsNaN (Math.Pow (             double.NaN, double.NegativeInfinity)), "#2");
 			Assert.IsTrue (double.IsNaN (Math.Pow (             double.NaN,                      -2)), "#2");
 			Assert.IsTrue (double.IsNaN (Math.Pow (             double.NaN,                      -1)), "#3");
-			Assert.IsTrue (double.IsNaN (Math.Pow (             double.NaN,                       0)), "#4");
+			Assert.IsFalse (double.IsNaN (Math.Pow (             double.NaN,                       0)), "#4");
```
Comment (luhenry) [inline mcs/class/corlib/Test/System/MathTest.cs]: Doesn't this break on top of .NET Framework?

Comment (EgorBo) [reply mcs/class/corlib/Test/System/MathTest.cs]: it does, but matches .NET Core and the IEEE 754 spec, see https://github.com/dotnet/coreclr/issues/9807

Comment (EgorBo) [reply mcs/class/corlib/Test/System/MathTest.cs]: or did you mean the case where we could run our tests on top of .NET Framework on Windows? AFAIR nobody maintains it and it doesn't work.

**Comment by marek-safar** [comment]
@monojenkins commit apidiff and csproj

**Comment by akoeplinger** [comment]
@EgorBo This caused two failures in acceptance-tests:

```
coreclr/tests/src/JIT/Directed/intrinsic/pow/pow1.exe:

MESSAGE:
x: 1, y: NaN, Pow(x,y): 1
FAILED
```

```
coreclr/tests/src/CoreMangLib/cti/system/math/mathpow.exe:

MESSAGE:
Beginning test case MathPow at 8/13/2018 5:21:32 PM
Random seed: 20010415; set environment variable CORECLR_SEED to this value to repro

[Positive]
Beginning scenario: PosTest1: Verify Pow when one of args is NaN .
Beginning scenario: PosTest2: Verify Pow when second arg is zero .
Beginning scenario: PosTest3: Verify Pow(NegativeInfinity, < 0) .
Beginning scenario: PosTest4: Verify Pow(NegativeInfinity, positive odd int) .
Beginning scenario: PosTest5: Verify Pow(NegativeInfinity, positive non-odd int) .
Beginning scenario: PosTest6: Verify Pow(< 0, non-int) .
Beginning scenario: PosTest7: Verify Pow(-1, NegativeInfinity) .
ERROR!!!-007.1: Return value is wrong!
Beginning scenario: PosTest8: Verify Pow(-1<x<1, NegativeInfinity) .
Beginning scenario: PosTest9: Verify Pow(-1<x<1, PositiveInfinity) .
Beginning scenario: PosTest10: Verify Pow(>1, NegativeInfinity) .
Beginning scenario: PosTest11: Verify Pow(>1, PositiveInfinity) .
Beginning scenario: PosTest12: Verify Pow( 0, < 0) .
Beginning scenario: PosTest13: Verify Pow( 0, > 0) .
Beginning scenario: PosTest14: Verify Pow( 1, any value) .
Beginning scenario: PosTest15: Verify Pow( PositiveInfinity, < 0) .
Beginning scenario: PosTest16: Verify Pow( PositiveInfinity, > 0) .

Ending test case at 8/13/2018 5:21:32 PM
FAIL

```

**Pull request merged**